airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLQueryBuilder.kt

```diff
@@ -0,0 +1,329 @@
+                NamedField(AIRBYTE_EXTRACTED_AT, FieldType(TimestampTypeWithoutTimezone, false)),
+                NamedField(AIRBYTE_META, FieldType(ObjectTypeWithoutSchema, false)),
+                NamedField(AIRBYTE_GENERATION_ID, FieldType(IntegerType, false)),
+            )
+
+        val airbyteFields = airbyteFinalTableFields.map { it.name }.toSet()
+
+        inline fun <reified T> deserialize(value: String): T =
+            Jsons.deserialize(value, T::class.java)
+    }
+
+    data class NamedField(val name: String, val type: FieldType)
+    data class NamedValue(val name: String, val value: AirbyteValue)
+
+    private val internalSchema: String = config.rawDataSchema
+    private val outputSchema: String = stream.descriptor.namespace ?: config.schema
+    private val tableName: String = stream.descriptor.name
+    private val fqTableName = "$outputSchema.$tableName"
+
+    val finalTableSchema: List<NamedField> =
+        airbyteFinalTableFields + extractFinalTableSchema(stream.schema)
+
+    fun createFinalTableIfNotExists(): String =
+        createTableIfNotExists(fqTableName, finalTableSchema)
+
+    fun createFinalSchemaIfNotExists(): String = createSchemaIfNotExists(outputSchema)
+
+    fun getFinalTableInsertColumnHeader(): String =
+        getFinalTableInsertColumnHeader(fqTableName, finalTableSchema)
+
+    fun populateStatement(
+        statement: PreparedStatement,
+        record: DestinationRecord,
+        schema: List<NamedField>
+    ) {
+        val toSqlType = AirbyteTypeToSqlType()
+        val recordObject = record.data as ObjectValue
+
+        var airbyteMetaStatementIndex: Int? = null
+        val airbyteMeta =
+            AirbyteRecordMessageMeta().apply {
+                changes = mutableListOf()
+                setAdditionalProperty("syncId", stream.syncId)
+            }
+        schema.forEachIndexed { index, field ->
+            val stmntIdx = index + 1
+            val value = recordObject.values[field.name]
+            val sqlType = toSqlType.convert(field.type.type)
+
+            if (value == null) {
+                if (field.name in airbyteFields) {
+                    when (field.name) {
+                        AIRBYTE_RAW_ID ->
+                            statement.setString(stmntIdx, UUID.randomUUID().toString())
+                        AIRBYTE_EXTRACTED_AT ->
+                            statement.setTimestamp(
+                                stmntIdx,
+                                Timestamp.from(Instant.ofEpochMilli(record.emittedAtMs))
+                            )
+                        AIRBYTE_GENERATION_ID -> statement.setLong(stmntIdx, stream.generationId)
+                        AIRBYTE_META -> airbyteMetaStatementIndex = stmntIdx
+                    }
+                } else {
+                    statement.setNull(stmntIdx, sqlType)
+                }
+            } else {
+                try {
+                    when (value) {
```

Comment:
It may make this cleaner to encapsulate this logic in an extension function on the value -- something like `value.setStatementValue(statement=statement)`.  The extension function would encapsulate this when block, making this code easier to test separate from the type conversion expectations.